src/jdk.jartool/share/classes/sun/tools/jar/Validator.java

```diff
@@ -62,20 +62,92 @@ final class Validator {
     private Set<String> concealedPkgs = Collections.emptySet();
     private ModuleDescriptor md;
     private String mdName;
+    private final ZipInputStream zis;
+    private final LinkedHashSet<String> cenEntryNames = new LinkedHashSet<>();
 
-    private Validator(Main main, ZipFile zf) {
+    private Validator(Main main, ZipFile zf, ZipInputStream zis) {
         this.main = main;
         this.zf = zf;
+        this.zis = zis;
         checkModuleDescriptor(MODULE_INFO);
     }
 
-    static boolean validate(Main main, ZipFile zf) throws IOException {
-        return new Validator(main, zf).validate();
+    static boolean validate(Main main, ZipFile zf, ZipInputStream zis) throws IOException {
+        return new Validator(main, zf, zis).validate();
+    }
+
+    // Detect duplicates when iterate through CEN entries
+    private void checkDuplicates(ZipEntry e) {
+        var cenEntryName = e.getName();
+        if (!cenEntryNames.add(cenEntryName)) {
+            warn(formatMsg("warn.validator.duplicate.cen.entry", cenEntryName));
+            isValid = false;
+        }
+    }
+
+    // Check if CEN contains the LOC entry by name
+    private boolean checkHasCenEntry(String locEntryName) {
```

Comment:
This method name and comment does not indicate a match is removed like the name `Next` in `checkNextCenEntry` implies - I suggest some way to indicate that a match is removed, like `checkUnorderedCenEntry`. `Has` implies a read-only operation that performs no modification.

Reply:
checkAnyCenEntry, Any vs Next make it relatively obvious on the difference.